lib/Dialect/FIRRTL/FIRRTLTypes.cpp

```diff
@@ -1826,6 +1939,81 @@ auto FEnumType::verify(function_ref<InFlightDiagnostic()> emitErrorFn,
   return success();
 }
 
+/// Return this type with any type aliases recursively removed from itself.
+FIRRTLBaseType FEnumType::getAnonymousType() {
+  if (this->isAnonymousType())
+    return *this;
+  return *this;
+}
+
+//===----------------------------------------------------------------------===//
+// FTypeAliasType
+//===----------------------------------------------------------------------===//
+
+struct circt::firrtl::detail::FTypeAliasStorage : mlir::TypeStorage {
+  using KeyTy = std::tuple<StringAttr, FIRRTLType>;
+
+  FTypeAliasStorage(StringAttr name, FIRRTLType innerType)
+      : name(name), innerType(innerType),
+        anonymousType(innerType.getAnonymousType()) {}
+
+  bool operator==(const KeyTy &key) const {
+    return key == KeyTy(name, innerType);
+  }
+
+  static llvm::hash_code hashKey(const KeyTy &key) {
+    return llvm::hash_combine(key);
+  }
+
+  static FTypeAliasStorage *construct(TypeStorageAllocator &allocator,
+                                      KeyTy key) {
+    return new (allocator.allocate<FTypeAliasStorage>())
+        FTypeAliasStorage(std::get<0>(key), std::get<1>(key));
+  }
+  StringAttr name;
+  FIRRTLType innerType;
+  FIRRTLType anonymousType;
+  FIRRTLBaseType passiveType;
+};
+
+auto FTypeAliasType::get(StringAttr name, FIRRTLType innerType)
+    -> FTypeAliasType {
+  return Base::get(name.getContext(), name, innerType);
+}
+
+auto FTypeAliasType::getName() const -> StringAttr { return getImpl()->name; }
+
+auto FTypeAliasType::getInnerType() const -> FIRRTLType {
+  return getImpl()->innerType;
+}
+
+FIRRTLBaseType FTypeAliasType::getPassiveType() {
+  auto *impl = getImpl();
+  if (impl->passiveType)
+    return impl->passiveType;
+  if (getRecursiveTypeProperties().isPassive)
+    return impl->passiveType = this->cast<FIRRTLBaseType>();
+  auto innerType = getInnerType().cast<FIRRTLBaseType>().getPassiveType();
+  return impl->passiveType =
```

Comment:
This cast will crash.

Furthermore `getPassiveType()` on a type alias ... surely (?) shouldn't return a type with same name but different contents?